validate-json.mjs

```diff
@@ -0,0 +1,225 @@
+const uint64Regex = /^[0-9]+$/;
+const int64Regex = /^-?[0-9]+$/;
+const hex128Regex = /^0[xX][0-9A-Fa-f]{1,32}$/;
+
+class State {
+  constructor() {
+    this.path = [];
+    this.errors = [];
+    this.warnings = [];
+  }
+
+  error(msg) {
+    this.errors.push({path: [...this.path], msg});
+  }
+
+  warn(msg) {
+    this.warnings.push({path: [...this.path], msg});
+  }
+
+  scope(scope, f) {
+    this.path.push(scope);
+    f();
+    this.path.pop();
+  }
+
+  result() {
+    return {errors: this.errors, warnings: this.warnings};
+  }
+
+  validate(obj, checks) {
+    Object.entries(checks).forEach(([key, check]) =>
+        this.scope(key, () => check(this, obj, key)));
+
+    Object.keys(obj).forEach(key => {
+      if (!(key in checks)) {
+        this.scope(key, () => this.warn('is an unknown field'));
```

Comment:
Suggestion:
"error type: field" (e.g. "unknown field: xxx") as suggested above in `formattedError` may be easier to parse and quicker to act upon than "field error type" ("xxx is an unknown field"). 
It also makes it potentially easier later to group errors per field, which maps better to the debugging flow (fix fieldX, then fix fieldY).

```suggestion
        this.scope(key, () => this.warn('unknown field'));
```